### ISSUE TITLE
Android - Permission Request Freezing and not resolving

### ISSUE THREAD

**Issue by FilipiRafael** [zoontek/react-native-permissions#902]
### Before submitting a new issue

- [X] I tested using the latest version of the library, as the bug might be already fixed.
- [X] I tested using a [supported version](https://github.com/reactwg/react-native-releases/blob/main/docs/support.md) of react native.
- [X] I checked for possible duplicate issues, with possible answers.

### Bug summary

The permission request for accessing the microphone in the requestAudioPermission function behaves unexpectedly when the user selects “Don’t Allow” twice. On the second attempt, the permission status is returned as “BLOCKED” but only once, after which the function freezes and does not resolve the promise. The await request() call halts execution, and the flow does not proceed. After reloading the app, the request resolves once but stops functioning on subsequent sessions, causing the permission flow to freeze again.

### Library version

4.0.4

### Environment info

```shell
System:
  OS: macOS 15.0.1
  CPU: (11) arm64 Apple M3 Pro
  Memory: 145.19 MB / 18.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 18.20.3
    path: ~/.nvm/versions/node/v18.20.3/bin/node
  Yarn:
    version: 1.22.22
    path: ~/.nvm/versions/node/v18.20.3/bin/yarn
  npm:
    version: 10.8.2
    path: ~/.nvm/versions/node/v18.20.3/bin/npm
  Watchman:
    version: 2024.05.06.00
    path: /opt/homebrew/bin/watchman
Managers:
  CocoaPods:
    version: 1.15.2
    path: /opt/homebrew/lib/ruby/gems/3.1.0/bin/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 24.0
      - iOS 18.0
      - macOS 15.0
      - tvOS 18.0
      - visionOS 2.0
      - watchOS 11.0
  Android SDK: Not Found
IDEs:
  Android Studio: 2024.1 AI-241.15989.150.2411.11948838
  Xcode:
    version: 16.0/16A242d
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 17.0.11
    path: /opt/homebrew/Cellar/openjdk@17/17.0.11/libexec/openjdk.jdk/Contents/Home/bin/javac
  Ruby:
    version: 3.1.6
    path: /opt/homebrew/opt/ruby@3.1/bin/ruby
npmPackages:
  "@react-native-community/cli": Not Found
  react:
    installed: 18.2.0
    wanted: 18.2.0
  react-native:
    installed: 0.74.5
    wanted: 0.74.5
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: false
iOS:
  hermesEnabled: true
  newArchEnabled: false
```

### Steps to reproduce

Set the Android Permission as "Ask every time", then try to record an audio. Don't allow twice times, the request will work twice times, then it will stop resolving. If you reload the app, it will resolve once and then stop again.

### Reproducible sample code

```js
const requestAudioPermission = async () => {
		try {
			const permission = Platform.OS === 'ios' ? PERMISSIONS.IOS.MICROPHONE : PERMISSIONS.ANDROID.RECORD_AUDIO;

			const checkResult = await check(permission);
			if (checkResult === RESULTS.GRANTED) {
				return true;
			}

			// it's freezing here, the promise doesn't resolve, it should return 'BLOCKED' or something at least
			const result = await request(permission);

			if (result === RESULTS.GRANTED) {
				return true;
			}

			setPermissionErrorType('microphone');
			return false;
		} catch (error) {
			setPermissionErrorType('microphone');
			return false;
		}
	};
```

**Comment by zoontek** [comment]
Tried it with latest `react-native-permissions` (`5.0.1`), eveything went well on a simulator or a real device (a Pixel 4a):

https://github.com/user-attachments/assets/4ceef974-47a7-4adb-89b6-a70062e45ec5

Please make sure your are using latest version.

**Comment by ned-tran-codeleap** [comment]
Hi @zoontek. 
I understand for demo video. I have faced the same situation and implement check method. But my case is requestMultiple, it does not resolve and freeze after 2 system dialogs popup. 
What do you think ?

**Comment by zoontek** [comment]
@NedTran Looks like https://github.com/zoontek/react-native-permissions/issues/966

**Comment by ned-tran-codeleap** [comment]
Hey @zoontek. WDYT?
This is my implementation for addressing the permission issue on Android.

Context

I encountered a “Not resolved” state when users denied the two system dialogs triggered by requestMultiple().
After the first two denials, the system dialog stops appearing entirely.

Implementation

```
public static async location(): Promise<void | boolean> {
  const checkSystemPermissions =
    await this.handleCheckSystemPermissionSetting([
      ActionLocation.UNIVERSAL_LOCATION,
    ]);

  if (checkSystemPermissions.length > 0) {
    this.handleOpenPermissionModalMultiple(checkSystemPermissions);
  } else {
    if (Platform.OS === 'ios') {
      return this._fnBatchImplIOS([
        {
          permissionId: PERMISSIONS.IOS.LOCATION_WHEN_IN_USE,
        },
      ]);
    } else if (Platform.OS === 'android') {
      return this._fnBatchImplAndroid([
        {
          permissionId: PermissionsAndroid.PERMISSIONS.ACCESS_COARSE_LOCATION,
          apiLevel: 23,
        },
        {
          permissionId: PermissionsAndroid.PERMISSIONS.ACCESS_FINE_LOCATION,
          apiLevel: 23,
        },
      ]);
    }
  }
}

```
```
private static async _fnBatchImplAndroid(
  batch: PermissionsBatch_ANDROID,
): Promise<void | boolean> {
  if (Platform.OS !== 'android') {
    return;
  }

  return new Promise((resolve, reject) => {
    PermissionsAndroid.requestMultiple(batch.map(i => i.permissionId)).then(
      (_result: { [key: string]: PermissionStatus }) => {
        const platformContants = (Platform as PlatformAndroidStatic).constants;
        const deniedPermissions: PermissionRejectError[] = [];

        const _fnCheckPermission = (permission: string): null | string => {
          if (_result[permission] === 'granted') {
            return null;
          }
          const rejectObject: PermissionRejectError = {
            os: Platform.OS,
            permissionType: this.mapPermissionToActionLocation(permission),
            reason: _result[permission],
            settingLevel: 'app',
            apiLevel: platformContants.Version,
          };
          deniedPermissions.push(rejectObject);
          return _result[permission];
        };

        for (const item of batch) {
          if (platformContants.Version >= item.apiLevel) {
            _fnCheckPermission(item.permissionId);
          }
        }

        if (deniedPermissions.length > 0) {
          reject(deniedPermissions); // Reject with all denied permissions
        } else {
          resolve(true);
        }
      },
    );
  });
}
```
Final Approach

After detecting that the user has denied the permission twice—and the system dialog will no longer appear—I switch to using check() to determine the permission state.
If the permission is still denied and the system dialog has already appeared twice, the app displays our custom permission modal instead.

This results in a significantly faster, more predictable, and more consistent permission flow.
